### DISCUSSION TITLE
Release pyodide as an npm/node package

### DISCUSSION THREAD

**Discussion by dmondev** [pyodide/pyodide#1403] [Ideas]
Are there any plans for releasing pyodide as an npm package ?

Being python for the web/browser, it makes sense to provide an npm package so it can be used along with the rest of the web development ecosystem, that revolves around node/npm and affiliates.

It seems that for the standard distribution, all we need is package.json + the contents of the build folder and of course an account in npmjs to publish (seems like there's another pyodide account there, that I'm not sure is related to you).

[Update] - 2021-04-02

I published an initial version based on 0.17.0a1 to the following org/package.

https://www.npmjs.com/package/@pyodide/pyodide

And also an example for running locally using a simple webpack-dev-server:
https://github.com/pyodude/pyodide-webpack

## Notes about this package

- This version includes the standard distribution minus scipy/dependencies plus plotly. As the plan for the core distribution changes (distribute only the minimal build), the npmjs package can easily reflect that.

- Had to remove scipy (and dependents), as the files are too big for github/npmjs. I think that for newer versions there shouldn't be the same  issue.

- I used 0.17.0a1 instead of dev(master), because of https://github.com/pyodide/pyodide/issues/1384. I'm mainly interested in using pyodide with plotly right now.

## About "@pyodide" in npmjs

I hope you forgive me for taking the **huge** liberty of publishing this under @pyodide/pyodide. The org name was available in npmjs, and I already saw a couple of pyodide packages in npmjs, so I didn't want you folks to "lose" the org, or have to go through the process of claiming it for you.

I'd be happy to give ownership to the npmjs project/org, just let me know. I can maintain it for the time being, as I'm using pyodide, and very interested in keeping the npmjs package up to date.

**Comment by hoodmane** [comment]
I think this is one of our top priorities for v0.18.0.

**Reply by dmondev** [reply]
I'm happy to help in any way I can. 
I created this to test (I published manually, but with "np" the process is pretty straightforward):
https://www.npmjs.com/package/@pyodude/pyodide

**Reply by hoodmane** [reply]
> I'm happy to help in any way I can.

Yeah, help with this is very welcome. I think the main blocker for a long time has been the fact that `src/pyodide.js` was a total mess. It's much better now, but there is still work to be done there. Last time I tried, adding `"use strict";` to the top still broke things.

We've almost completely eliminated the global variables, but we still need `pyodide` to be a global variable for `loadPackage`. I think this will be necessary for as long as we are supporting browsers that can't use ES6 modules in webworkers (see #1364).

**Reply by rth** [reply]
@dmondev I think we are going to move forward on this. Could you please give us admin rights to @pyodide/pyodide. Usernames are `ryurchak` and `hoodmane`? 

If I'm not mistaken @hoodmane you also have access to https://www.npmjs.com/package/pyodide, right? So which package do we make the official one in the end? I understand that having an org namespace is useful, particularly if we are going to explore option 2b in https://github.com/pyodide/pyodide/issues/14#issuecomment-875116949 one day, but at the same time `npm install pyodide` seems much easier and more readable than `npm install @pyodide/pyodide`

**Reply by rth** [reply]
And @dmondev feedback on https://github.com/pyodide/pyodide/issues/14#issuecomment-875116949 and generally what we currently have in https://github.com/pyodide/pyodide/tree/main/src/js with respect to what you would expect to see in the npm package would be very welcome.

**Reply by hoodmane** [reply]
> If I'm not mistaken @hoodmane you also have access to npmjs.com/package/pyodide, right?

Yes that is correct.

**Reply by dmondev** [reply]
Sorry, @rth , I was offline for a while. I just invited you to the pyodide org at npmjs. I'll transfer ownership to you as soon as you join.

**Reply by dmondev** [reply]
@rth , you're now owner of @pyodide/pyodide, I added @hoodmane as well. Please remove me from the organization, I don't need to be there :)

**Reply by rth** [reply]
Thanks a lot @dmondev !

**Comment by subwaymatch** [comment]
Just want to say huge thanks to @dmondev @rrth @hoodmane! Packaging pyodide into an npm package will make it much more accessible.

**Comment by subwaymatch** [comment]
@rth @hoodmane Thank you for all the work in releasing v0.18! Follow-up question. Is the current `pyodide` npm package only for node.js, not the browser? The example in https://www.npmjs.com/package/pyodide is for Node.js only and I was wondering if I can use it in React (Next.js) projects.

**Reply by hoodmane** [reply]
You can use it in the same way in browser (e.g., with rollup or some other bundler). In that case you should set `indexURL` equal to `https://cdn.jsdelivr.net/pyodide/v0.18.0/full/`.

**Reply by subwaymatch** [reply]
I'll give it a try with next.js. Thank you!
